core/src/main/scala/org/ensime/core/javac/JavaDocFinding.scala

```diff
@@ -0,0 +1,129 @@
+        }
+      }
+    }.getOrElse(List())
+  }
+
+  private def memberSig(info: CompilationInfo, m: MethodInvocationTree,
+    candidates: List[ExecutableElement], name: String): Option[String] = {
+    val args = m.getArguments
+    val byArity = candidates.filter { c => c.getParameters.length == args.length }
+    val withTypeDeltas = byArity.map { c => typeDeltas(info, m, c) }.toArray
+    val compatible = withTypeDeltas.filter { case (c, deltas) => !deltas.contains(Int.MaxValue) }
+    val sorted = compatible.sortWith { (c1, c2) => c1._2.sum < c2._2.sum }
+    sorted.headOption.map {
+      case (c, deltas) =>
+        val args = c.getParameters.map { p => fqn(info, p.asType).map(_.mkString).getOrElse("NA") }.mkString(",")
+        s"$name($args)"
+    }
+  }
+
+  private def typeMirror(info: CompilationInfo, t: Tree): Option[TypeMirror] = {
+    Option(info.getTrees().getTypeMirror(info.getTrees().getPath(info.getCompilationUnit(), t)))
+  }
+
+  private def typeElement(info: CompilationInfo, t: Tree): Option[Element] = {
+    typeMirror(info, t).map(info.getTypes().asElement)
+  }
+
+  private def fqn(info: CompilationInfo, p: TreePath): Option[DocFqn] = {
+    val tm = Option(info.getTrees().getTypeMirror(p))
+    tm.flatMap { tm => fqn(info, tm) }
+  }
+
+  private def fqn(info: CompilationInfo, t: Tree): Option[DocFqn] = {
+    Option(info.getTrees().getPath(info.getCompilationUnit(), t)).flatMap { p => fqn(info, p) }
+  }
+
+  private def fqn(info: CompilationInfo, tm: TypeMirror): Option[DocFqn] = {
+
+    def parseFqn(qualifiedName: String) = {
+      val (front, back) = qualifiedName.split("\\.").partition { s => s.forall(Character.isLowerCase) }
+      DocFqn(front.mkString("."), back.mkString("."))
+    }
+
+    // "Using instanceof is not necessarily a reliable idiom for
+    // determining the effective class of an object in this modeling
+    // hierarchy since an implementation may choose to have a single
+    // object implement multiple TypeMirror subinterfaces." --
+    // TypeMirror docs
+    tm match {
+      case tm: DeclaredType if tm.getKind == TypeKind.DECLARED => {
+        tm.asElement match {
+          case te: TypeElement => Some(parseFqn(te.getQualifiedName.toString))
+          case _ => {
+            None
+          }
+        }
+      }
+      case tm: PrimitiveType if tm.getKind.isPrimitive => Some(DocFqn("", tm.toString))
+      case _ => None
+    }
+  }
+
+  // For each (argument, parameter) pair compute an integer distance between them.
+  // Int.MaxValue denotes an incompatible pairing (e.g. reference argument supplied for
+  // primitive param). 0 denotes type equality. 
+  private def typeDeltas(info: CompilationInfo, m: MethodInvocationTree, c: ExecutableElement): (ExecutableElement, Array[Int]) = {
+    val types = info.getTypes()
+    val args = m.getArguments
+    val params = c.getParameters
+    var i = 0
+    val result = new Array[Int](args.length)
+    while (i < args.length) {
```

Comment:
I don't think we need a mutable `i` here, `for { i <- 0 to args.length}` should suffice. I assume `Array` is necessary to meet a Java API.